### ISSUE TITLE
document dirty state not preserved on reload

### ISSUE THREAD

**Issue by jjallaire** [rstudio/rstudio#6846]
### System details

    RStudio Edition : Desktop and Server (server in development mode)
    RStudio Version :  1.4 (daily)
    OS Version      :  10.15.4 
    R Version       : 3.6.1

### Steps to reproduce the problem

1. Start with an empty workspace (no source editor tabs)
2. Open an R script and make some edits (don't save, leave the document dirty).
3. Open another R script, make some edits, then save it.
4. Leaving things be for a few seconds (so the IDE persists the active tab)
5. Reload the browser (right click and select Reload from the context menu)
6. Dirty state of first document is shown as "clean" (no asterisk or red text)

### Describe the problem in detail

This issue can cause data loss, as if the dirty state isn't updated you may think that all of your work in a tab has been saved to the filesystem (when it is in fact still just in the source database).

### Describe the behavior you expected

Dirty state is correctly update.

- [ x] I have read the guide to submitting good bug reports at https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports . 
- [ x] I have installed the latest version of RStudio and confirmed that the issue still persists.
- [ x] If I am reporting a RStudio crash, I have included a diagnostics report. https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report
- [ x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by jmcphers** [comment]
This is a 1.3 bug if it reproduces there (to be verified)

**Comment by jmcphers** [comment]
Possibly related to **https://github.com/rstudio/rstudio/issues/6824**?

**Comment by dfalty** [comment]
I was able to reproduce this on 1.3.957-4 and 1.4.324-4 but not on 1.2.5042-1 in my RHEL 7.6 RSP environment so it does look like a regression.

**Comment by dfalty** [comment]
It looks like you don't even have to mess with another file for me to reproduce. (RSP 1.3.957-1)

1. Create session
2. Open an already existing file
3. Make a change and wait a couple of seconds
4. Refresh the browser (or go to RSP homepage and back)
5. Observe "clean" state shown on file

**Comment by jmcphers** [comment]
@kevinushey Could you take a look? I tracked this down to a problem wherein `diffImpl` returns a non-empty diff even when given identical original and new values. This causes us to send the server a "diff" that doesn't change the content, which removes the dirty flag.

**Comment by jmcphers** [comment]
As a proof of concept, adding this to the top of `diffImpl` fixes the bug by ensuring an empty diff is generated when the contents are equal (probably not the right change)

```javascript
      if (origVal === newVal)
      {
         return {
            "replacement": "",
            "offset": 0,
            "length": 0
         };
      }
```

**Comment by jmcphers** [comment]
Related: https://github.com/rstudio/rstudio/pull/6680, https://github.com/rstudio/rstudio/issues/6676, https://github.com/rstudio/rstudio/pull/5615

**Comment by dfalty** [comment]
Closing https://github.com/rstudio/rstudio/pull/6853

**Comment by R-Hannibal** [comment]
QA Notes:

The fix for this ticket changed the codepath used for document saves (thanks @kevinushey for the info). While we do not expect to see further regressions or issues in this area, it may be worth triple checking as part of the fix verification, especially considering how far into 1.3 QA is:

- After saving a document, what you see in the editor indeed matches what is on disk;

- If you have an unsaved buffer, ensure that the state of that buffer is preserved after a browser reload (relevant for RStudio Server).

**Comment by dfalty** [comment]
Verified fixed on RSP 1.3.958-2 and Mac Desktop 1.3.958. Did some testing around different unsaved buffer states and making sure the editor matches up with what's on disk.
